File: Bank_Management_System.cpp
#include <iostream>
#include <string>

using namespace std;

// Function prototypes
void createAccount(string &name, int &accountNumber, double &balance);
void depositMoney(double &balance);
void withdrawMoney(double &balance);
void checkBalance(const double &balance);
void displayAccountDetails(const string &name, const int &accountNumber, const double &balance);

//SubFunctions that will be called throughout the program
void createAccount(string &name, int &accountNumber, double &balance) {
    cout << "Enter your name: ";
    cin.ignore(); // Clear the newline character from the buffer
    getline(cin, name);
    
    cout << "Enter your account number: ";
    cin >> accountNumber;

    do {
        cout << "Enter your initial deposit (must be greater than 0): R";
        cin >> balance;
        if (balance <= 0) {
            cout << "Initial deposit must be greater than 0. Please try again." << endl;
        }
    } while (balance <= 0);

    cout << "Account created successfully!" << endl;
}

void depositMoney(double &balance) {
    double deposit;
    cout << "Enter amount to deposit: R";
    cin >> deposit;

    if (deposit > 0) {
        balance += deposit;
        cout << "Deposit successful! New balance: R" << balance << endl;
    } else {
        cout << "Deposit amount must be greater than 0." << endl;
    }
}

void withdrawMoney(double &balance) {
    double withdrawal;
    cout << "Enter amount to withdraw: R";
    cin >> withdrawal;

    if (withdrawal > 0 && withdrawal <= balance) {
        balance -= withdrawal;
        cout << "Withdrawal successful! New balance: R" << balance << endl;
    } else if (withdrawal > balance) {
        cout << "Withdrawal amount exceeds current balance!" << endl;
    } else {
        cout << "Withdrawal amount must be greater than 0." << endl;
    }
}

void checkBalance(const double &balance) {
    cout << "Current balance: R" << balance << endl;
}

void displayAccountDetails(const string &name, const int &accountNumber, const double &balance) {
    cout << "--- Account Details ---" << endl;
    cout << "Account Holder: " << name << endl;
    cout << "Account Number: " << accountNumber << endl;
    cout << "Current Balance: R" << balance << endl;
}

//Main funtion that entails the whole logic of the program
int main() {
    string name;
    int accountNumber = 0;
    double balance = 0.0;
    int choice;

    do {
        // Display the main menu
        cout << "--- Bank Account Management System ---" << endl;
        cout << "1. Create Account" << endl;
        cout << "2. Deposit Money" << endl;
        cout << "3. Withdraw Money" << endl;
        cout << "4. Check Balance" << endl;
        cout << "5. Display Account Details" << endl;
        cout << "6. Exit" << endl;
        cout << "Enter your choice (1-6): ";
        cin >> choice;

        switch (choice) {
            case 1:
                createAccount(name, accountNumber, balance);
                break;
            case 2:
                depositMoney(balance);
                break;
            case 3:
                withdrawMoney(balance);
                break;
            case 4:
                checkBalance(balance);
                break;
            case 5:
                displayAccountDetails(name, accountNumber, balance);
                break;
            case 6:
                cout << "Exiting the system. Goodbye!" << endl;
                break;
            default:
                cout << "Invalid choice! Please select a valid option." << endl;
        }
    } while (choice != 6);

    return 0;
}

File: Bank_management_System.py
class BankAccount:
    def __init__(self, name, account_number, initial_deposit):
        self.name = name
        self.account_number = account_number
        self.balance = initial_deposit

    def deposit(self, amount):
        if amount <= 0:
            print("Deposit amount must be greater than 0.")
            return
        self.balance += amount
        print(f"Deposited R{amount:.2f}. New balance is R{self.balance:.2f}.")

    def withdraw(self, amount):
        if amount <= 0:
            print("Withdrawal amount must be greater than 0.")
            return
        if amount > self.balance:
            print("Insufficient funds for this withdrawal.")
            return
        self.balance -= amount
        print(f"Withdrew R{amount:.2f}. New balance is R{self.balance:.2f}.")

    def check_balance(self):
        print(f"Current balance: R{self.balance:.2f}")

    def display_account_details(self):
        print(f"--- Account Details ---")
        print(f"Account Holder: {self.name}")
        print(f"Account Number: {self.account_number}")
        print(f"Current Balance: R{self.balance:.2f}")


def create_account():
    name = input("Enter your name: ")
    account_number = input("Enter your account number: ")
    initial_deposit = float(input("Enter your initial deposit (must be greater than 0): "))
    while initial_deposit <= 0:
        initial_deposit = float(input("Invalid amount. Enter your initial deposit (must be greater than 0): "))
    return BankAccount(name, account_number, initial_deposit)


def main():
    print("--- Bank Account Management System ---")
    account = None

    while True:
        print("\n1. Create Account")
        print("2. Deposit Money")
        print("3. Withdraw Money")
        print("4. Check Balance")
        print("5. Display Account Details")
        print("6. Exit")
        choice = input("Enter your choice (1-6): ")

        if choice == '1':
            account = create_account()
            print("Account created successfully!")

        elif choice == '2':
            if account:
                amount = float(input("Enter amount to deposit: "))
                account.deposit(amount)
            else:
                print("No account found. Please create an account first.")

        elif choice == '3':
            if account:
                amount = float(input("Enter amount to withdraw: "))
                account.withdraw(amount)
            else:
                print("No account found. Please create an account first.")

        elif choice == '4':
            if account:
                account.check_balance()
            else:
                print("No account found. Please create an account first.")

        elif choice == '5':
            if account:
                account.display_account_details()
            else:
                print("No account found. Please create an account first.")

        elif choice == '6':
            print("Exiting the system. Goodbye!")
            break

        else:
            print("Invalid choice. Please select a valid option.")


if __name__ == "__main__":
    main()
